### ISSUE TITLE
[Task 29] REINFORCE++ / REINFORCE++-baseline implementation

### ISSUE THREAD

**Issue by Marvex2022** [redai-infra/Relax#160]
Related to #86, Task 29.

### 1. 代码仓库、基线 commit、开发分支

- 代码仓库：`redai-infra/Relax`
- 基线分支：`main`
- 基线 commit：`8a54679e971087566bfda939a5f75649e07fb861`
- 开发分支：`feat/task29-reinforce-plus-plus`
- 贡献者：`@Marvex2022`

本任务将在上述固定基线下，补齐并验证 REINFORCE++ 与REINFORCE++-baseline 的算法配置、return/advantage、loss、监控指标、数值测试和训练对比。

#### 软件环境

- Docker 镜像：`ghcr.io/redai-infra/relaxrl:latest`
- Python：`3.12.3`
- PyTorch：`2.11.0+cu129`
- CUDA Runtime：`12.9`
- NVIDIA Driver 支持的 CUDA 版本：`13.1`
- 训练精度：BF16
- 训练后端：Megatron-LM
- Rollout 后端：SGLang
- 调度框架：Ray

#### 硬件环境

- CPU：`Intel(R) Xeon(R) Platinum 8558`
- 内存：`2.0 TiB`
- Swap：`8.0 GiB`
- 可用 GPU：`8 × NVIDIA H100 80GB`

### 3. 启动脚本及完整命令

计划新增参数化脚本：

```text
examples/algorithms/run-qwen3-0.6B-reinforce-plus-plus.sh
```

通过 `ADVANTAGE_ESTIMATOR` 选择算法，不复制多份仅算法名称不同的recipe。

#### REINFORCE++

```bash
ADVANTAGE_ESTIMATOR=reinforce_plus_plus \
SEED=1234 \
CUDA_VISIBLE_DEVICES=0 \
bash examples/algorithms/run-qwen3-0.6B-reinforce-plus-plus.sh
```

#### REINFORCE++-baseline

```bash
ADVANTAGE_ESTIMATOR=reinforce_plus_plus_baseline \
SEED=1234 \
CUDA_VISIBLE_DEVICES=1 \
bash examples/algorithms/run-qwen3-0.6B-reinforce-plus-plus.sh
```

#### GRPO 对比基线

```bash
ADVANTAGE_ESTIMATOR=grpo \
SEED=1234 \
CUDA_VISIBLE_DEVICES=2 \
bash examples/algorithms/run-qwen3-0.6B-reinforce-plus-plus.sh
```

脚本计划执行以下完整训练命令：

```bash
python3 -m relax.entrypoints.train \
  --resource '{"actor": [1, 1], "rollout": [1, 1]}' \
  --max-staleness 0 \
  --num-data-storage-units 1 \
  --colocate \
  --use-health-check \
  --swiglu \
  --num-layers 28 \
  --hidden-size 1024 \
  --ffn-hidden-size 3072 \
  --num-attention-heads 16 \
  --group-query-attention \
  --num-query-groups 8 \
  --use-rotary-position-embeddings \
  --disable-bias-linear \
  --normalization RMSNorm \
  --norm-epsilon 1e-6 \
  --rotary-base 1000000 \
  --vocab-size 151936 \
  --kv-channels 128 \
  --qk-layernorm \
  --hf-checkpoint /your/model/Qwen3-0.6B \
  --ref-load /your/model/Qwen3-0.6B \
  --megatron-to-hf-mode bridge \
  --warm-hf-checkpoint-page-cache \
  --prompt-data /your/data/gsm8k/train.jsonl \
  --input-key question \
  --label-key answer \
  --apply-chat-template \
  --rollout-shuffle \
  --rm-type math \
  --num-rollout 200 \
  --rollout-batch-size 4 \
  --n-samples-per-prompt 8 \
  --rollout-max-response-len 2048 \
  --rollout-temperature 1.0 \
  --global-batch-size 16 \
  --balance-data \
  --use-fault-tolerance \
  --skip-eval-before-train \
  --eval-interval 20 \
  --eval-prompt-data aime /your/data/aime-2024/aime-2024.jsonl \
  --eval-input-key prompt \
  --eval-label-key label \
  --n-samples-per-eval-prompt 4 \
  --eval-max-response-len 2048 \
  --log-passrate \
  --optimizer adam \
  --lr 1e-6 \
  --lr-decay-style constant \
  --weight-decay 0.1 \
  --adam-beta1 0.9 \
  --adam-beta2 0.98 \
  --advantage-estimator "${ADVANTAGE_ESTIMATOR}" \
  --normalize-advantages \
  --gamma 1.0 \
  --use-kl-loss \
  --kl-loss-coef 0.00 \
  --kl-loss-type low_var_kl \
  --eps-clip 0.2 \
  --use-rollout-logprobs \
  --use-metrics-service \
  --tb-project-name Relax/dev/task29 \
  --tb-experiment-name \
    "qwen3-0.6b-${ADVANTAGE_ESTIMATOR}-seed-${SEED}" \
  --tensor-model-parallel-size 1 \
  --pipeline-model-parallel-size 1 \
  --context-parallel-size 1 \
  --expert-model-parallel-size 1 \
  --expert-tensor-parallel-size 1 \
  --calculate-per-token-loss \
  --use-dynamic-batch-size \
  --max-tokens-per-gpu 8192 \
  --log-probs-max-tokens-per-gpu 8192 \
  --rollout-num-gpus-per-engine 1 \
  --sglang-mem-fraction-static 0.45 \
  --attention-dropout 0.0 \
  --hidden-dropout 0.0 \
  --accumulate-allreduce-grads-in-fp32 \
  --attention-softmax-in-fp32 \
  --attention-backend flash \
  --seed "${SEED}"
```

本任务重点新增或修改的算法参数为：

```text
--advantage-estimator reinforce_plus_plus
--advantage-estimator reinforce_plus_plus_baseline
--normalize-advantages
--gamma
```

正式对比实验保持模型、数据、训练步数、batch、采样参数、学习率、response 长度和硬件一致，仅改变 advantage estimator 及其必要配置。

#### 预计相关文件

- `relax/utils/training/ppo_utils.py`
  - `get_reinforce_plus_plus_returns`
  - `get_reinforce_plus_plus_baseline_advantages`
  - 相关 mask、normalization 和 policy loss 计算

- `relax/components/advantages.py`
  - 两种 estimator 的 advantage 分发与统计

- `relax/backends/megatron/loss.py`
  - return/advantage 接入
  - mask 与 loss reduction
  - 分布式统计和训练指标

- `relax/core/registry.py`
  - 注册并区分两个算法变体

- `relax/utils/arguments.py`
  - 参数名称、默认值、约束和错误提示

- `examples/algorithms/`
  - 参数化训练 recipe

- `tests/utils/training/`
- `tests/components/`
- `tests/backends/megatron/`
- `tests/core/`
  - 独立参考实现、参数、注册、分布式和 loss 测试

- 算法设计文档
  - 明确公式、归一化维度、mask 和 reduction 口径
  - 对比 REINFORCE++、baseline、GRPO、GSPO 和 SAPO

#### 改动范围

- 补齐 REINFORCE++ 与 baseline 版本的配置和注册；
- 明确并实现两种算法的 return/advantage 口径；
- 对齐 Megatron loss 和 Advantages 组件中的计算行为；
- 增加必要的训练监控指标；
- 增加独立 reference 数值测试和分布式统计测试；
- 增加一个共享的参数化 recipe；
- 保证现有算法默认行为不受影响。

#### 明确不做

- 不实现 fully-async REINFORCE++；
- 不引入 critic 或 value model；
- 不实现 RLOO、Dr.GRPO 或其他额外算法；
- 不修改 reward 函数和数据集格式；
- 不修改模型结构、SGLang 或 checkpoint 格式；
- 不进行与任务无关的大规模代码重构；
- 不通过复制两份 recipe 代替两个算法变体的真实接入。

#### 基线数据

- 模型：`Qwen3-0.6B`
- 训练集：`GSM8K train`
- 评测集：`AIME-2024`
- Reward：Relax math rule-based reward
- 对比算法：`GRPO`
- 训练步数：每种算法 `200 steps`
- Samples per prompt：`8`
- 随机种子：`42、1234、2026`
- 单次实验：`1 × H100 80GB`

计划比较：

| 算法 | Steps | Seeds |
|---|---:|---:|
| GRPO | 200 | 3 |
| REINFORCE++ | 200 | 3 |
| REINFORCE++-baseline | 200 | 3 |

#### 数值验收指标

- 关键 return、advantage 和 loss 张量与独立 PyTorch reference
  逐元素对齐；
- 数值误差目标：

```text
torch.allclose(actual, expected, atol=1e-6, rtol=1e-6)
```

- 测试覆盖：
  - 变长 response；
  - padding 和有效 token mask；
  - 全零 reward；
  - 单样本；
  - baseline group；
  - 零方差；
  - 分布式统计；
  - loss reduction；
  - 参数注册与校验。

#### 训练指标

报告以下指标：

- reward mean/std；
- policy loss；
- PPO KL；
- clip fraction；
- advantage mean/std；
- response length；
- pass@1 / pass@4；
- step time；
- tokens/s 或 samples/s；
- GPU 峰值显存。

#### 统计口径

- 三种算法采用相同代码基线、模型、数据、硬件和训练预算；
- 每个算法运行 3 个相同种子集合；
- 报告各 seed 的完整曲线；
- 汇总 3 个 seed 的均值和标准差；
- 训练稳定性统计使用最后 20 个 step；
- padding 和 prompt token 不参与有效 response token 统计；
- 分布式统计结果应与单进程拼接后的 reference 一致。

#### 目标值

- 所有目标单元测试通过；
- reference 对齐误差不超过 `1e-6`；
- 三种算法均完成 200 个 training step；
- 不出现未解释的 NaN/Inf、静默丢样本或错误 mask；
- 两个算法名称、日志和配置可以明确区分；
- 现有 GRPO、GSPO、SAPO 等路径不回归；
- 不要求 REINFORCE++ 必须优于 GRPO，但要求同预算结果可复现、
  训练过程稳定且差异可解释。

### 6. 交付方式

Proposal Issue → 算法实现 PR → 数值测试与训练报告。
